# Microdados de Vacinação - OpenDataSUS

Este documento possui o código que estou utilizando para processar o arquivo de microdados da vacinação disponibilizado pelo SUS. As informações sobre a base pode ser encontrada [neste link](https://opendatasus.saude.gov.br/dataset/covid-19-vacinacao). Existem duas formas de acessar os dados: (1) a partir do arquivo CSV, que atualmente possui mais de 1GB de dados ou (2) através de uma API Elasticsearch. 

Estou usando uma versão disponibilizada pelo [Brasil.io](brasil.io). Detalhes deste documento podem ser encontrados nesse status do twitter: https://twitter.com/brasil_io/status/1361080270791864321. Os dados processados pelo brasil.io estão mais limpos, consistentes e com alterações que permitiram gerar um arquivo menor. O processamento a seguir vai ser feito em cima deste arquivo que pode ser acessado a partir do status do brasil.io. O arquivo não está disponível neste repositório. 

Minha proposta neste documento é mostrar como podemos usar o [pandas](https://pandas.pydata.org/) para olhar um pouco mais para esses dados. Quando esses dados estiverem mais consolidados, devo usa-los para criar visualizações que irão complementar o painel da covid que mantenho no endereço https://covid19.data2learning.com.  

**Vale ressaltar que estes dados contém informações do Brasil inteiro e que inconsistências podem existir por diversos motivos como erros na hora da coletar e da alimentação dos dados.**

O compartilhamento destes dados são de extrema importância para que todos possam ter accesso às informações completas dos registros da vacinação. Até o dia 20/02 os dados ainda estavam incompletos totalizando pouco mais de 5,5 milhões de registros (primeira e segunda doses). Outras fontes informavam nesta mesma data um quantitativo em torno de 5,8 milhões só da primeira dose.

A medida que o arquivo for sendo atualizado, esse doumento também será. 

In [1]:
# Bibliotecas utilizadas 

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Carregando os dados a partir do arquivo microdados_vacinacao.csv

data_microdados = pd.read_csv("data/microdados_vacinacao.csv")

In [3]:
data_microdados.head()

paciente_codigo_grupo                               vacina vacina_lote  \
0                    9.0  Covid-19-Coronavac-Sinovac/Butantan      210015   
1                    9.0         Vacina Covid-19 - Covishield    4120Z005   
2                    2.0         Vacina Covid-19 - Covishield    4120Z005   
3                    2.0  Covid-19-Coronavac-Sinovac/Butantan      210012   
4                    9.0  Covid-19-Coronavac-Sinovac/Butantan   202009015   

             estabelecimento_razao_social  \
0    PREFEITURA DO MUNICIPIO DE SAO PAULO   
1              CENTRAL DE IMUNOBIOLOGICOS   
2    MANAUS SECRETARIA MUNICIPAL DE SAUDE   
3                      SMS DE CAMPO MAIOR   
4  PREFEITURA MUNICIPAL DE PRES VENCESLAU   

                                     estabelecimento  paciente_cep  \
0              HOSP MUN DR IGNACIO PROENCA DE GOUVEA        2083.0   
1                REDE DE FRIO DO MUNICIPIO DE MARICA       24900.0   
2                            USF MJ PM SALVIO BELOTA       69095.0   
3                                CENTRO DE SAUDE III       64280.0   
4  CENTRO DE SAUDE I TACITO LEITE DE CARVALHO E S...       19380.0   

  estabelecimento_unidade_federativa          sistema_origem  \
0                                 SP  Pendente identificação   
1                                 RJ                Novo PNI   
2                                 AM  Pendente identificação   
3                                 PI                Novo PNI   
4                                 SP  Pendente identificação   

    paciente_municipio paciente_etnia  ... data_aplicacao  \
0            São Paulo            NaN  ...     2021-02-15   
1               Maricá            NaN  ...     2021-02-04   
2               Manaus        Amarela  ...     2021-02-02   
3          Campo Maior        Amarela  ...     2021-02-12   
4  Ribeirão dos Índios         Branca  ...     2021-02-12   

                         documento_uuid  paciente_sexo_biologico  \
0  bbcda716-4acd-5f60-9b9b-7ce55971419c                        F   
1  3b48177e-03cd-50de-baf7-59eb4854d194                        F   
2  a987dca4-1419-5b9d-a84e-eec57ea45581                        F   
3  5d4cf2a2-5525-5662-b3e7-0bb70b3900fb                        M   
4  b99fb876-a5cc-5ae1-86c5-99dacf859bc2                        F   

   estabelecimento_codigo_cnes paciente_unidade_federativa  \
0                      2084473                          SP   
1                       127744                          RJ   
2                      2013886                          AM   
3                      2369850                          PI   
4                      2051249                          SP   

   paciente_codigo_pais  paciente_pais paciente_codigo_ibge_municipio  \
0                  10.0         Brasil                      3550308.0   
1                  10.0         Brasil                      3302700.0   
2                  10.0         Brasil                      1302603.0   
3                  10.0         Brasil                      2202208.0   
4                  10.0         Brasil                      3543238.0   

            paciente_subgrupo paciente_codigo_etnia  
0                      Outros                  99.0  
1               Enfermeiro(a)                  99.0  
2  Pessoas de 80 anos ou mais                   4.0  
3  Pessoas de 80 anos ou mais                   4.0  
4                      Outros                   1.0  

[5 rows x 29 columns]

In [4]:
data_microdados.columns

Index(['paciente_codigo_grupo', 'vacina', 'vacina_lote',
       'estabelecimento_razao_social', 'estabelecimento', 'paciente_cep',
       'estabelecimento_unidade_federativa', 'sistema_origem',
       'paciente_municipio', 'paciente_etnia', 'estabelecimento_municipio',
       'paciente_idade', 'paciente_codigo_subgrupo', 'numero_dose',
       'paciente_nacionalidade', 'codigo_vacina',
       'estabelecimento_codigo_ibge_municipio', 'paciente_uuid',
       'paciente_grupo', 'data_aplicacao', 'documento_uuid',
       'paciente_sexo_biologico', 'estabelecimento_codigo_cnes',
       'paciente_unidade_federativa', 'paciente_codigo_pais', 'paciente_pais',
       'paciente_codigo_ibge_municipio', 'paciente_subgrupo',
       'paciente_codigo_etnia'],
      dtype='object')

In [5]:
data_microdados.dtypes

paciente_codigo_grupo                    float64
vacina                                    object
vacina_lote                               object
estabelecimento_razao_social              object
estabelecimento                           object
paciente_cep                             float64
estabelecimento_unidade_federativa        object
sistema_origem                            object
paciente_municipio                        object
paciente_etnia                            object
estabelecimento_municipio                 object
paciente_idade                           float64
paciente_codigo_subgrupo                   int64
numero_dose                                int64
paciente_nacionalidade                    object
codigo_vacina                              int64
estabelecimento_codigo_ibge_municipio      int64
paciente_uuid                             object
paciente_grupo                            object
data_aplicacao                            object
documento_uuid      

In [6]:
data_microdados.shape

(5335315, 29)

## Olhando mais para os dados

Nesse tipo de dado, que é gerado a partir de informações coletadas no Brasil inteiro por diversos municípios, pode acontecer de ter algumas inconsistências. Muitas das vezes essas inconsistências se dar por conta de erros no processo de digitação e coleta destas informações. Mesmo sem ter como saber a origem do problema, vale a pena verificar algumas possíveis inconsistências para um eventual pré-processamento da base e utilização da mesma para construção de dashboards.

### Datas 

A primeira coisa que vou fazer é olhar a quantidade de registros de vacinação por dia. 

In [7]:
data_microdados.groupby(by=['data_aplicacao','numero_dose']).count()[['paciente_uuid']]

paciente_uuid
data_aplicacao numero_dose               
2020-01-20     1                        1
2020-01-28     1                        1
2020-12-20     1                        1
2020-12-28     1                        2
2020-12-29     1                        1
...                                   ...
2021-02-17     2                   106636
2021-02-18     1                    63773
               2                   116475
2021-02-19     1                    40092
               2                    65488

[81 rows x 1 columns]

Existem registros anteriores à data oficial de início da vacinação. Vamos ver quantos registros de vacinação antes do dia 17/01 quando a primeira dose da vacina foi aplicada: https://noticias.uol.com.br/saude/ultimas-noticias/redacao/2021/01/17/enfermeira-de-sp-negra-moradora-de-itaquera-e-1-vacinada-no-brasil.htm.

In [8]:
data_microdados[data_microdados['data_aplicacao'] < '2021-01-17'].groupby(by=['data_aplicacao','numero_dose']).count()[['paciente_uuid']]

paciente_uuid
data_aplicacao numero_dose               
2020-01-20     1                        1
2020-01-28     1                        1
2020-12-20     1                        1
2020-12-28     1                        2
2020-12-29     1                        1
2021-01-01     1                        2
2021-01-03     1                        1
2021-01-04     1                        1
2021-01-11     1                        5
               2                        1
2021-01-12     1                        1
2021-01-15     1                        1
2021-01-16     1                        1

Observe que no total são apenas 12 registros com datas anteriores à 17/01.

### Idade

Uma outra verificação é em relação a idade de quem recebeu as doses. Para isso vamos olhar o quantitativo no campo `paciente_idade_grupo` que é criado a partir do campo original `paciente_idade`. O campo `paciente_idade_grupo` discretiza os dados em faixas faixas hetárias. Foi utilizada a seguinte classificação de idades: Criança (0-12 anos), adolescente (13-18 anos), adulto (19-59 anos) e idoso (acima de 60 anos).

In [9]:
data_microdados['paciente_idade_grupo'] = pd.cut(data_microdados['paciente_idade'],bins=[0,13,19,60,float("inf")],right=False,labels=['criança','adolescente','adulto','idoso'])
data_microdados[['paciente_idade_grupo']].value_counts()

paciente_idade_grupo
adulto                  3674113
idoso                   1642758
adolescente               18084
criança                     359
dtype: int64

Somando o quantitativo de **crianças** e **adolescentes** tempos no total 18443 registros de pessoas menores que 17 anos que receberam a vacina. Sabendo que a aplicação da vacina está sendo em grupos prioritários que não incluem crianças e adolescentes, esses dados podem ser resultado de alguma inconsistência. Vamos ver esse quantitativo por idade. 

In [10]:
data_microdados[data_microdados['paciente_idade'] < 18].groupby(by='paciente_idade').count()[['paciente_uuid']]

paciente_uuid
paciente_idade               
0.0                        23
1.0                        21
2.0                        19
3.0                        21
4.0                        30
5.0                        25
6.0                        20
7.0                        29
8.0                        25
9.0                        31
10.0                       38
11.0                       28
12.0                       49
13.0                       42
14.0                       52
15.0                      107
16.0                      350
17.0                     1277

Vamos relacionar esses dados com o campo `paciente_subgrupo`:

In [11]:
temp_data = data_microdados[data_microdados['paciente_idade'] < 18].groupby(by=['paciente_idade','paciente_subgrupo']).count()[['paciente_uuid']]
initial_age = 0
print('Idade 0')
for index, row in temp_data.iterrows():
        if index[0] != initial_age:
            print()
            print('Idade %i' % index[0])
            initial_age = index[0]
        print(' %i %s' % (row['paciente_uuid'], index[1]))

Idade 0
 2 Assistente Social
 9 Outros
 1 Pessoal da Limpeza
 1 Pessoas de 70 a 74 anos
 2 Pessoas de 80 anos ou mais
 7 Povos indígenas em terras indígenas
 1 Técnico de Enfermagem

Idade 1
 2 Enfermeiro(a)
 4 Outros
 2 Pessoas de 60 nos ou mais Institucionalizadas
 7 Pessoas de 80 anos ou mais
 5 Povos indígenas em terras indígenas
 1 Técnico de Enfermagem

Idade 2
 1 Cuidador de Idosos
 2 Enfermeiro(a)
 6 Outros
 1 Pessoas de 70 a 74 anos
 3 Pessoas de 80 anos ou mais
 5 Povos indígenas em terras indígenas
 1 Recepcionista

Idade 3
 1 Médico
 10 Outros
 1 Pessoas de 60 nos ou mais Institucionalizadas
 1 Pessoas de 75 a 79 anos
 2 Pessoas de 80 anos ou mais
 5 Povos indígenas em terras indígenas
 1 Segurança

Idade 4
 2 Enfermeiro(a)
 1 Fisioterapeutas
 1 Fonoaudiólogo
 2 Médico
 10 Outros
 2 Pessoas de 70 a 74 anos
 3 Pessoas de 80 anos ou mais
 6 Povos indígenas em terras indígenas
 1 Profissionais de Educação Física
 1 Recepcionista
 1 Técnico de Enfermagem

Idade 5
 1 Motorista d

Veja que pelos dados temos médicos, enfermeiros, motoristas com idades inferiores à 18 anos. 

### Recebimento das doses

O campo `paciente_id` identifica unicamente cada uma das pessoas que receberam a vacina. Com essa informação é possível fazer algumas verificações a respeito da duplicidade de informação. Vamos dar uma olhada nisso em vários aspectos. 

Primeiro vamos ver se existe dados duplicados entre os pacientes que receberam a primeira dose. 

In [12]:
all_dose1 = data_microdados[data_microdados['numero_dose'] == 1]

Rodando o código a seguir, é possível ver que existem pouco mais de 37000 registros duplicados.

In [13]:
all_dose1[all_dose1[['paciente_uuid']].duplicated()]

paciente_codigo_grupo                               vacina  \
17467                      9.0         Vacina Covid-19 - Covishield   
28379                      9.0  Covid-19-Coronavac-Sinovac/Butantan   
36289                      2.0         Vacina Covid-19 - Covishield   
41695                      9.0  Covid-19-Coronavac-Sinovac/Butantan   
41826                      9.0  Covid-19-Coronavac-Sinovac/Butantan   
...                        ...                                  ...   
5335078                    9.0         Vacina Covid-19 - Covishield   
5335114                    9.0         Vacina Covid-19 - Covishield   
5335196                    9.0  Covid-19-Coronavac-Sinovac/Butantan   
5335267                    9.0  Covid-19-Coronavac-Sinovac/Butantan   
5335289                    9.0  Covid-19-Coronavac-Sinovac/Butantan   

        vacina_lote                       estabelecimento_razao_social  \
17467      4120Z005             PREFEITURA MUNICIPAL DE JARAGUA DO SUL   
28379     202010020                               MUNICIPIO DE PELOTAS   
36289      4120Z005                      PREFEITURA MUNICIPAL DE ANORI   
41695    0202010034             PREFEITURA MUNICIPAL DE JARAGUA DO SUL   
41826    0202010034             PREFEITURA MUNICIPAL DE JARAGUA DO SUL   
...             ...                                                ...   
5335078    4120Z005             PREFEITURA MUNICIPAL DE JARAGUA DO SUL   
5335114    4120Z005                          PREFEITURA DE SANTO ANDRE   
5335196   202010027  IRMANDADE DA SANTA CASA DE MISERICORDIA DE SAO...   
5335267   202010038            PREFEITURA MUNICIPAL DE OLIVEIRA FORTES   
5335289   202010029                PREFEITURA MUNICIPAL DE SANTO ANDRE   

                                           estabelecimento  paciente_cep  \
17467                                CENTRAL DE IMUNIZACAO       89270.0   
28379        CENTRAL REDE DE FRIO MUNICIPL NUCLEO MUNCIPAL       96080.0   
36289                 POSTO DE SAUDE FAUSTO GOMES DA ROCHA           NaN   
41695                                CENTRAL DE IMUNIZACAO       89253.0   
41826                                CENTRAL DE IMUNIZACAO       89254.0   
...                                                    ...           ...   
5335078                              CENTRAL DE IMUNIZACAO       89251.0   
5335114   HOSPITAL DE CAMPANHA COVID 19 PEDRO DELL ANTONIA        9240.0   
5335196                  HOSPITAL NOSSA SENHORA DA POMPEIA       44360.0   
5335267         UNIDADE BASICA DE SAUDE DE OLIVEIRA FORTES       36250.0   
5335289  CENTRO HOSPITALAR DE SANTO ANDRE DR NEWTON DA ...        8280.0   

        estabelecimento_unidade_federativa          sistema_origem  \
17467                                   SC  Pendente identificação   
28379                                   RS                Novo PNI   
36289                                   AM                Novo PNI   
41695                                   SC  Pendente identificação   
41826                                   SC  Pendente identificação   
...                                    ...                     ...   
5335078                                 SC  Pendente identificação   
5335114                                 SP  Pendente identificação   
5335196                                 BA                Novo PNI   
5335267                                 MG                Novo PNI   
5335289                                 SP  Pendente identificação   

        paciente_municipio paciente_etnia  ...  \
17467           Guaramirim         Branca  ...   
28379              Pelotas         Branca  ...   
36289                Anori         Branca  ...   
41695       Jaraguá do Sul         Branca  ...   
41826       Jaraguá do Sul         Branca  ...   
...                    ...            ...  ...   
5335078     Jaraguá do Sul         Branca  ...   
5335114        Santo André            NaN  ...   
5335196          São Félix        Amarela  ...   
5335267    Ol

O código acima exibe somente as linhas duplicadas. Podemos ver os registros duplicados um seguido do outro com o código a seguir. 

In [14]:
all_dose1[all_dose1[['paciente_uuid']].duplicated(keep=False)].sort_values(by='paciente_uuid')

paciente_codigo_grupo                               vacina  \
3335773                    3.0  Covid-19-Coronavac-Sinovac/Butantan   
3268559                    3.0  Covid-19-Coronavac-Sinovac/Butantan   
4605956                    9.0  Covid-19-Coronavac-Sinovac/Butantan   
3952578                    9.0  Covid-19-Coronavac-Sinovac/Butantan   
3930212                    9.0  Covid-19-Coronavac-Sinovac/Butantan   
...                        ...                                  ...   
1111526                    9.0  Covid-19-Coronavac-Sinovac/Butantan   
659811                     2.0  Covid-19-Coronavac-Sinovac/Butantan   
2604405                    2.0  Covid-19-Coronavac-Sinovac/Butantan   
2464439                    9.0  Covid-19-Coronavac-Sinovac/Butantan   
2977437                    9.0  Covid-19-Coronavac-Sinovac/Butantan   

        vacina_lote        estabelecimento_razao_social  \
3335773   202010038  PREFEITURA MUNICIPAL DE ESMERALDAS   
3268559   202010038  PREFEITURA MUNICIPAL DE ESMERALDAS   
4605956   202009014      PREFEITURA DA CIDADE DO RECIFE   
3952578   202009014      PREFEITURA DA CIDADE DO RECIFE   
3930212   202010020   MUNICIPIO DE IJUI PODER EXECUTIVO   
...             ...                                 ...   
1111526   202010027              MUNICIPIO DE BARREIRAS   
659811    202010031    PREFEITURA MUNICIPAL DE ANAPOLIS   
2604405   202010031    PREFEITURA MUNICIPAL DE ANAPOLIS   
2464439   202010019   MS INCA HC I HOSPITAL DO CANCER I   
2977437   202010019   MS INCA HC I HOSPITAL DO CANCER I   

                                    estabelecimento  paciente_cep  \
3335773                           CAPS 2 ESMERALDAS           NaN   
3268559                           CAPS 2 ESMERALDAS           NaN   
4605956             US 162 POLICLINICA ALBERT SABIN           NaN   
3952578             US 162 POLICLINICA ALBERT SABIN           NaN   
3930212          ESTRATEGIA SAUDE DA FAMILIA ESF 18       98700.0   
...                                             ...           ...   
1111526   COPIM COORDENACAO MUNICIPAL DE IMUNIZACAO       47800.0   
659811   CENTRAL MUNICIPAL DA REDE DE FRIO ANAPOLIS           NaN   
2604405  CENTRAL MUNICIPAL DA REDE DE FRIO ANAPOLIS           NaN   
2464439                MS INCA HOSPITAL DO CANCER I           NaN   
2977437                MS INCA HOSPITAL DO CANCER I           NaN   

        estabelecimento_unidade_federativa sistema_origem paciente_municipio  \
3335773                                 MG       Novo PNI          Sem-Peixe   
3268559                                 MG       Novo PNI          Sem-Peixe   
4605956                                 PE       Novo PNI             Recife   
3952578                                 PE       Novo PNI             Recife   
3930212                                 RS       Novo PNI               Ijuí   
...                                    ...            ...                ...   
1111526                                 BA       Novo PNI          Barreiras   
659811                                  GO       Novo PNI           Anápolis   
2604405                                 GO       Novo PNI           Anápolis   
2464439                                 RJ       Novo PNI     Rio de Janeiro   
2977437                                 RJ       Novo PNI     Rio de Janeiro   

        paciente_etnia  ...                        documento_uuid  \
3335773          Preta  ...  39c87f38-9fa5-5525-9809-269859091a0b   
3268559          Preta  ...  63515f1a-52e1-5bb2-9ec5-0108697703d0   
4605956            NaN  ...  a5ab4518-7c63-5b80-9e50-4a84844b50a3   
3952578            NaN  ...  a62ae5ed-01ec-5f56-92b9-b234c1a321f0   
3930212         Branca  ...  7c9b77ad-59ad-53a2-9cab-baee6e3abb22   
...                ...  ...                                   ...   
1111526         Branca  ...  9c4ef8d8-92bc-512c-8440-72a9d7e2e26d   
659811             NaN  ...  896bc39e-1b07-5ae9-a2ed-17e9e05c6216   
2604405            NaN  ...  25a1830

Vamos fazer a mesma coisa com a segunda dose:

In [15]:
all_dose2 = data_microdados[data_microdados['numero_dose'] == 2]
all_dose2[all_dose2[['paciente_uuid']].duplicated()]

paciente_codigo_grupo                               vacina  \
29006                      9.0  Covid-19-Coronavac-Sinovac/Butantan   
38082                      9.0  Covid-19-Coronavac-Sinovac/Butantan   
44808                      9.0  Covid-19-Coronavac-Sinovac/Butantan   
89015                      3.0  Covid-19-Coronavac-Sinovac/Butantan   
109373                     9.0  Covid-19-Coronavac-Sinovac/Butantan   
...                        ...                                  ...   
5319434                    7.0  Covid-19-Coronavac-Sinovac/Butantan   
5319475                    9.0  Covid-19-Coronavac-Sinovac/Butantan   
5320194                    9.0  Covid-19-Coronavac-Sinovac/Butantan   
5322544                    3.0  Covid-19-Coronavac-Sinovac/Butantan   
5322687                    9.0  Covid-19-Coronavac-Sinovac/Butantan   

        vacina_lote              estabelecimento_razao_social  \
29006     202010020      PREFEITURA MUNICIPAL DE TRES DE MAIO   
38082        210010                      MUNICIPIO DE LAJEADO   
44808        210015    PREFEITURA MUNICIPAL DE JARAGUA DO SUL   
89015       210011A         SMS RIO CMS DR ALBERT SABIN AP 21   
109373       210015    PREFEITURA MUNICIPAL DE JARAGUA DO SUL   
...             ...                                       ...   
5319434   202010034                       MINISTERIO DA SAUDE   
5319475      125589  SECRETARIA MUNICIPAL DE SAUDE DE ARACAJU   
5320194      200279    SECRETARIA MUNICIPAL DE SAUDE DE NATAL   
5322544   202010031         PREFEITURA MUNICIPAL DE CAPELINHA   
5322687   202009011           PREFEITURA MUNICIPAL DE MOSSORO   

                                     estabelecimento  paciente_cep  \
29006                       UNIDADE CENTRAL DE SAUDE       98910.0   
38082      ESTRATEGIA DE SAUDE DA FAMILIA MONTANHA 1           NaN   
44808                          CENTRAL DE IMUNIZACAO       89260.0   
89015                  SMS CMS DR ALBERT SABIN AP 21       22451.0   
109373                         CENTRAL DE IMUNIZACAO       89260.0   
...                                              ...           ...   
5319434    UNIDADE DE SAUDE INDIGENA JENIPAPO KANIDE       61700.0   
5319475     US ONESIMO PINTO FILHO JARDIM CENTENARIO       49220.0   
5320194  SECRETARIA MUNICIPAL DE SAUDE NIVEL CENTRAL       59570.0   
5322544       CENTRO DE SAUDE JOSE PIMENTA CAPELINHA       39680.0   
5322687                 UPA RAIMUNDO BENJAMIM FRANCO       59604.0   

        estabelecimento_unidade_federativa  \
29006                                   RS   
38082                                   RS   
44808                                   SC   
89015                                   RJ   
109373                                  SC   
...                                    ...   
5319434                                 CE   
5319475                                 SE   
5320194                                 RN   
5322544                                 MG   
5322687                                 RN   

                                            sistema_origem paciente_municipio  \
29006                                             Novo PNI       Três de Maio   
38082                               Pendente identificação            Estrela   
44808                               Pendente identificação     Jaraguá do Sul   
89015                                             Novo PNI     Rio de Janeiro   
109373                              Pendente identificação     Jaraguá do Sul   
...                                                    ...                ...   
5319434                                           Novo PNI            Aquiraz   
5319475  Sistema utilizado pela secretaria para registr...              Arauá   
5320194                                          RN+Vacina        Ceará-Mirim   
5322544                                           Novo PNI          Capelinha   
5322687                                          RN+Vacina            Mossoró   

        

Um pouco mais de 3600 linhas duplicas. Exibindo os registros duplicados no mesmo dataframe, temos:

In [16]:
all_dose2[all_dose2[['paciente_uuid']].duplicated(keep=False)].sort_values(by='paciente_uuid')

paciente_codigo_grupo                               vacina  \
554069                     7.0  Covid-19-Coronavac-Sinovac/Butantan   
341869                     7.0  Covid-19-Coronavac-Sinovac/Butantan   
634356                     9.0  Covid-19-Coronavac-Sinovac/Butantan   
3779278                    9.0  Covid-19-Coronavac-Sinovac/Butantan   
2732466                    7.0  Covid-19-Coronavac-Sinovac/Butantan   
...                        ...                                  ...   
3683651                    9.0  Covid-19-Coronavac-Sinovac/Butantan   
755474                     9.0  Covid-19-Coronavac-Sinovac/Butantan   
2048734                    9.0  Covid-19-Coronavac-Sinovac/Butantan   
1730367                    9.0  Covid-19-Coronavac-Sinovac/Butantan   
563382                     9.0  Covid-19-Coronavac-Sinovac/Butantan   

        vacina_lote                       estabelecimento_razao_social  \
554069    202010025  DISTRITO SANITARIO ESPECIAL DE SAUDE INDIGENA ...   
341869    202010025  DISTRITO SANITARIO ESPECIAL DE SAUDE INDIGENA ...   
634356    202010031  PREFEITURA MUNICIPAL DE SAO GONCALO DO RIO ABAIXO   
3779278   202010031  PREFEITURA MUNICIPAL DE SAO GONCALO DO RIO ABAIXO   
2732466   202010025  DISTRITO SANITARIO ESPECIAL DE SAUDE INDIGENA ...   
...             ...                                                ...   
3683651   202010018                    PREFEITURA MUNICIPAL DE SARANDI   
755474    202010018                    PREFEITURA MUNICIPAL DE SARANDI   
2048734   202010018                    PREFEITURA MUNICIPAL DE SARANDI   
1730367   202010034         SECRETARIA MUNICIPAL DA SAUDE DE JOINVILLE   
563382    202010034         SECRETARIA MUNICIPAL DA SAUDE DE JOINVILLE   

                                      estabelecimento  paciente_cep  \
554069                              POLO BASE FEIJOAL       69630.0   
341869                              POLO BASE FEIJOAL       69630.0   
634356   PRONTO ATENDIMENTO SAO GONCALO DO RIO ABAIXO       35935.0   
3779278  PRONTO ATENDIMENTO SAO GONCALO DO RIO ABAIXO       35935.0   
2732466                             POLO BASE FEIJOAL       69630.0   
...                                               ...           ...   
3683651                UNIDADE BASICA DE SAUDE AURORA       85935.0   
755474                 UNIDADE BASICA DE SAUDE AURORA       85935.0   
2048734                UNIDADE BASICA DE SAUDE AURORA       85935.0   
1730367                         SAE UNIDADE SANITARIA       88717.0   
563382                          SAE UNIDADE SANITARIA       88717.0   

        estabelecimento_unidade_federativa sistema_origem  \
554069                                  AM       Novo PNI   
341869                                  AM       Novo PNI   
634356                                  MG       Novo PNI   
3779278                                 MG       Novo PNI   
2732466                                 AM       Novo PNI   
...                                    ...            ...   
3683651                                 PR       Novo PNI   
755474                                  PR       Novo PNI   
2048734                                 PR       Novo PNI   
1730367                                 SC       Novo PNI   
563382                                  SC       Novo PNI   

                paciente_municipio paciente_etnia  ...  \
554069           Benjamin Constant       Indígena  ...   
341869           Benjamin Constant       Indígena  ...   
634356   São Gonçalo do Rio Abaixo          Parda  ...   
3779278  São Gonçalo do Rio Abaixo          Parda  ...   
2732466          Benjamin Constant            NaN  ...   
...                            ...            ...  ...   
3683651        Assis Chateaubriand            NaN  ...   
755474         Assis Chateaubriand            NaN  ...   
2048734        Assis Chateaubriand            NaN  ...   
1730367                     Sangão         Branca  ...   
563382                      Sangão         

Uma outra verificação é se existem registros de segunda dose sem a primeira. Vamos utilizar o merge dos dois dataframes `all_dose1` e `all_dose2`. Para simplificar, a visualização, vamos usar somente as informações `paciente_id` e `numero_dose`. Vou fazer um merge das duas bases utilizando a base da dose2 para o join. Vou usar um Right Join, ou seja, vou selecionar os registros da dose 2 e fazer um merge com os registros da dose 1. Em seguida selecionar aqueles nos quais a a dose 1 está nula. 

In [49]:
selected_dose1 = all_dose1[['paciente_uuid','numero_dose']]
selected_dose2 = all_dose2[['paciente_uuid','numero_dose']]
selected_merge = pd.merge(selected_dose1, selected_dose2, how ='right', suffixes=['1','2'], on =['paciente_uuid']) 
selected_merge = selected_merge.drop_duplicates()
selected_merge

paciente_uuid  numero_dose1  numero_dose2
0       6e42a649-2b79-5f41-91d0-b892ee35a974           1.0             2
1       1e286f7a-7b47-55cd-ae4d-a2db02d6b1bb           1.0             2
2       1256214b-d2eb-5aa3-b167-c01bb9107431           1.0             2
3       0e7af0eb-c951-5206-b989-ae4e6ee6bf04           1.0             2
4       78b42db1-1e32-5db8-84fc-4918ca99b08e           1.0             2
...                                      ...           ...           ...
867441  1199832b-28d8-557e-8f3a-29b88d4bbaf2           1.0             2
867442  ac6c90ff-9715-57f5-99cb-0309ae2014a6           1.0             2
867443  ed7081aa-75ce-5812-8b66-8551965dfbe9           1.0             2
867444  122c6132-db57-58f1-8af7-eb42bdfa0b3d           1.0             2
867445  cbb80b48-524c-57bf-bb1d-7e10cc58e822           1.0             2

[846870 rows x 3 columns]

In [50]:
selected_merge[selected_merge['numero_dose1'].isnull()]

paciente_uuid  numero_dose1  numero_dose2
35      207ca278-5587-5a50-bcc0-3ba7a3af2e94           NaN             2
73      e2717ffa-69bd-5bca-8d56-74431370ddcf           NaN             2
74      3e7b19b5-5bb4-5316-8931-e57ca55c1781           NaN             2
84      0788c0de-7f83-56bd-8641-20b94a1496f0           NaN             2
119     47d151f7-e64b-5e04-ac61-377714bd4d4f           NaN             2
...                                      ...           ...           ...
867259  afadfebf-4ec0-56b0-9b2d-0ed339626a4c           NaN             2
867263  62ff8875-d787-51c2-a7a9-ea874c356148           NaN             2
867373  c5553537-5c11-5757-bc51-266b4b8566b4           NaN             2
867377  d7bcb6b1-ea6f-57aa-8be8-e5f50d12aa6f           NaN             2
867402  b31f81ac-b67d-58d6-ac14-e745feff8b15           NaN             2

[39320 rows x 3 columns]

Foram encontrados 39320 registros que estão marcados como segunda dose mas  não tem a primeira dose correspondente. 

## Extraindo os totais de cada variável categórica 

Uma outra coisa que ajuda a enxergar um pouco como esses dados estão distribuídos é verificar o quantitativo por raça, sexto, idade, grupos. Vamos exibir essas informações utilizando o `value_counts` que será aplicado às variáveis categóricas. Vou fazer isso apenas para a primeira dose. Para isso, vou usar o campo `numero_dose`. A informação de UF será utilizada a do estabelecimento para garantir o quantitativo por estado, já que a UF do paciente não é necessariamente a de onde ele tomou a vacina. Essa informação está no campo `estabelecimento_unidade_federativa`.

In [19]:
data_microdados.columns

Index(['paciente_codigo_grupo', 'vacina', 'vacina_lote',
       'estabelecimento_razao_social', 'estabelecimento', 'paciente_cep',
       'estabelecimento_unidade_federativa', 'sistema_origem',
       'paciente_municipio', 'paciente_etnia', 'estabelecimento_municipio',
       'paciente_idade', 'paciente_codigo_subgrupo', 'numero_dose',
       'paciente_nacionalidade', 'codigo_vacina',
       'estabelecimento_codigo_ibge_municipio', 'paciente_uuid',
       'paciente_grupo', 'data_aplicacao', 'documento_uuid',
       'paciente_sexo_biologico', 'estabelecimento_codigo_cnes',
       'paciente_unidade_federativa', 'paciente_codigo_pais', 'paciente_pais',
       'paciente_codigo_ibge_municipio', 'paciente_subgrupo',
       'paciente_codigo_etnia', 'paciente_idade_grupo'],
      dtype='object')

In [20]:
cat_columns = ['paciente_idade_grupo','paciente_etnia','estabelecimento_unidade_federativa','paciente_sexo_biologico','paciente_grupo','paciente_subgrupo']
data_selected = data_microdados[data_microdados['numero_dose'] == 1]
dt_ = {}

for c in cat_columns:
    dt_[c] = pd.DataFrame(data_selected[c].value_counts())
    dt_[c].reset_index(inplace=True)
    dt_[c].columns = ['info','total']
    


In [21]:
dt_['paciente_sexo_biologico']

info    total
0    F  3047043
1    M  1437705
2    I        3

In [22]:
dt_['paciente_grupo']

info    total
0                            Trabalhadores de Saúde  2918013
1                                      Faixa Etária  1226490
2                                   Povos Indígenas   181259
3    Pessoas de 60 anos ou mais institucionalizadas   122398
4                           Pessoas com Deficiência    12920
5                                      Comorbidades    10603
6                  Povos e Comunidades Tradicionais     6966
7                  Forças de Segurança e Salvamento     3505
8                         Trabalhadores da Educação     1721
9                   Forças Armadas (membros ativos)      315
10                       Pessoas em Situação de Rua      200
11  Funcionário do Sistema de Privação de Liberdade      189
12                   População Privada de Liberdade       62
13                         Trabalhadores Portuários       52
14                        Trabalhadores Industriais       36

In [23]:
for index, row in dt_['paciente_subgrupo'].iterrows():
    print("%s: %i" % (row['info'], row['total']))

Outros: 1676197
Pessoas de 80 anos ou mais: 1024365
Técnico de Enfermagem: 330761
Povos indígenas em terras indígenas: 181259
Enfermeiro(a): 160611
Médico: 149856
Pessoas de 60 nos ou mais Institucionalizadas: 122398
Pessoal da Limpeza: 96891
Recepcionista: 84244
Pessoas de 75 a 79 anos: 64981
Odontologista: 64200
Motorista de Ambulância: 56163
Pessoas de 70 a 74 anos: 54279
Pessoas de 60 a 64 anos: 49811
Fisioterapeutas: 39119
Farmacêutico: 36039
Pessoas de 65 a 69 anos: 33054
Psicólogo: 27351
Cuidador de Idosos: 26902
Auxiliar de Enfermagem: 25303
Cozinheiro e Auxiliares: 24106
Técnico de Odontologia: 20070
Segurança: 17828
Assistente Social: 16263
Nutricionista: 14650
Pessoas com Deficiência Institucionalizadas: 10367
Estudante: 9757
Biomédico: 8580
Profissionais de Educação Física: 7663
Fonoaudiólogo: 7070
Quilombola: 6965
Funcionário do Sistema Funerário c/ cadáveres potencialmente contaminados: 6749
Médico Veterinário: 5852
Hipertensão de difícil controle ou com complicações/lesã

In [24]:
dt_['paciente_idade_grupo']

info    total
0       adulto  2930673
1        idoso  1537709
2  adolescente    16055
3      criança      314

In [25]:
dt_['paciente_etnia']

info    total
0    Branca  1751519
1     Parda   756040
2   Amarela   500496
3     Preta   163999
4  Indígena   129703

In [26]:
dt_['estabelecimento_unidade_federativa']

info    total
0    SP  1537972
1    MG   357248
2    RS   296016
3    RJ   265605
4    BA   198941
5    AM   176356
6    PE   167934
7    PR   154582
8    GO   142169
9    MA   117615
10   PA   114568
11   CE   101821
12   PB    97930
13   SC    97173
14   ES    95710
15   DF    80972
16   RN    80763
17   MT    78959
18   MS    63017
19   AL    57853
20   PI    53331
21   TO    34569
22   RO    34487
23   SE    34483
24   AC    16488
25   AP    14197
26   RR    13993

## Pré-processamento e Visualização de dados

_Em breve_